### ISSUE TITLE
Multi-character sandwiches

### ISSUE THREAD

**Issue by maxfl** [machakann/vim-sandwich#1]
Sorry, it is not clear from documentation. Is it possible to define multi-character sandwiches or is such a feature planned?

**Comment by machakann** [comment]
Yes, it is possible.
```vim
let g:sandwich#recipes = deepcopy(g:sandwich#default_recipes)
let g:sandwich#recipes += [{'buns': ['```', '```']}]
```
Adding the lines into your vimrc, you can use the pattern for triple backquotes. Is it matched with what you want?

If you could tell a  specified situation (language, syntax...), probably I could suggest a better example :)

**Comment by maxfl** [comment]
Thank you. That's nice. There is a problem of a mapping clash with '`' though. 

So far I'm interested in the following surroundings python/text/latex:
1. Python triple quotes: '''|''' and """|""".
2. Unicode quotes: “|”, „|“, «|».
3. Latex quotes: `|', ``|'', "`|"', <<|>> or their command representations
4. Latex command/environment definition and some others: \command{|}, \begin{env}|\end{env}, \begingroup|\endgroup. (This item may be out of the vim-sandwich scope though).

The latex quote definition as well as python triple quotes will clash with ', " quotes. Is there a way to set a custom hotkey for the sandwich recipe?

**Comment by maxfl** [comment]
Just realized that there are three documentation files and one of them has a description of recipes and it seem to have a custom hotkey. I will try it.

**Comment by maxfl** [comment]
So far I've tried the following:
```vim
        let g:sandwich#recipes = [
        \	{'buns': ['“', '”'],   'expand_range': 0, 'input': [ '1' ] },
        \	{'buns': ['„', '“'],   'expand_range': 0, 'input': [ '2' ] },
        \	{'buns': ['«', '»'],   'expand_range': 0, 'input': [ '3' ] },
        \	{'buns': ["``", "''"], 'expand_range': 0, 'input': [ '4' ], 'quoteescape': 1 }, 
        \	{'buns': ['<<', '>>'], 'expand_range': 0, 'input': [ '5' ] },
        \ ]
        let g:sandwich#recipes = deepcopy(g:sandwich#default_recipes)+g:sandwich#recipes
```
The sandwicd-add works as expected, vib works as well, but I can not make use of replace/delete functions by sd1/sd2,sr12 etc. Am I still missing something?

**Comment by machakann** [comment]
Hi, it seems you've already found some resolutions by yourself, but just for examples I put.

1. python triple quote
    ```vim
    let g:sandwich#recipes += [
          \   {'buns': ["'''", "'''"], 'filetype': ['python'], 'nesting': 0, 'input': ["3'"]},
          \   {'buns': ['"""', '"""'], 'filetype': ['python'], 'nesting': 0, 'input': ['3"']},
          \ ]
    ```
    'input': ['3"'] might not be good example, somehow bothering my finger. Make hotkeys as you like.

2. Unicode quotes
    Your example seems good.

3. Latex quotes
    ```vim
    let g:sandwich#recipes += [
        \   {'buns': ['`', "'"],    'filetype': ['tex', 'plaintex'], 'nesting': 0, 'input': ["`'"]},
        \   {'buns': ['``', "''"],  'filetype': ['tex', 'plaintex'], 'nesting': 0, 'input': ['`"']},
        \   {'buns': ['"`', '"'''], 'filetype': ['tex', 'plaintex'], 'nesting': 0, 'input': ['"`']},
        \   {'buns': ['<<', '>>'],  'filetype': ['tex', 'plaintex'], 'nesting': 0, 'input': ['<<']},
        \ ]
    ```

4. Latex command/environment definition and some others
    I have not written tex so long time, please check it works or not. Excuse me!
    `\command{|}` and `\begingroup | \endgroup` can be used like this.
    ```vim
    let g:sandwich#recipes += [
        \   {'buns': ['\command{', '}'], 'filetype': ['tex', 'plaintex'], 'nesting': 0, 'input': ['\co']},
        \   {'buns': ['\begingroup ', '\endgroup'], 'filetype': ['tex', 'plaintex'], 'motionwise': ['char', 'block'], 'nesting': 0, 'input': ['\gr']},
        \   {'buns': ['\begingroup',  '\endgroup'], 'filetype': ['tex', 'plaintex'], 'motionwise': ['line'],          'nesting': 0, 'input': ['\gr']},
        \ ]
    ```

    `\begin{env} | \end{env}` is actually difficult because {env} also need to be input.
    Just to let you know, 'expr' option can be used. (sorry, this option is called as 'eval' in old version. If it won't work please update.)

    ```vim
    let g:TexEnvironments = [
        \   'array', 'center', 'description', 'enumerate', 'eqnarray', 'equation',
        \   'equation*', 'figure', 'flushleft', 'flushright', 'itemize', 'list',
        \   'minipage', 'picture', 'quotation', 'quote', 'tabbing', 'table',
        \   'tabular', 'tabular*', 'thebibliography', 'theorem', 'titlepage',
        \   'verbatim', 'verse'
        \ ]

    function! TexEnvCompl(argread, cmdline, cursorpos) abort
        let n = strlen(a:argread)
        let list = copy(g:TexEnvironments)
        if n > 0
            let list = filter(list, 'v:val[: n-1] ==# a:argread')
        endif
        return list
    endfunction

    function! TexEnvInput(is_head) abort
        if a:is_head
            let g:TexEnvLast = input('Environment-name: ', '', 'customlist,TexEnvCompl')
            let command = printf('\begin{%s}', g:TexEnvLast)
        else
            let command = printf('\end{%s}', g:TexEnvLast)
        endif
        return command
    endfunction

    let g:sandwich#recipes += [
        \   {
        \       'buns'    : ['TexEnvInput(1)', 'TexEnvInput(0)'],
        \       'filetype': ['tex', 'plaintex'],
        \       'expr'    : 1,
        \       'nesting' : 1,
        \       'input'   : ['\env']
        \   },
        \ ]
    ```

    For example type `saiw\env` then the cursor moves to commandline and prompt to input {env}. Type environment-name and press <Enter> then a word would be surrounded. Environment-names can be completed by <Tab> key, but not comprehensive only the items in `g:TexEnvironments` can be suggested.
    However it works, it might be better to use other snippet plugins for this use...

**Comment by machakann** [comment]
> So far I've tried the following:
>
>         let g:sandwich#recipes = [
>         \   {'buns': ['“', '”'],   'expand_range': 0, 'input': [ '1' ] },
>         \   {'buns': ['„', '“'],   'expand_range': 0, 'input': [ '2' ] },
>         \   {'buns': ['«', '»'],   'expand_range': 0, 'input': [ '3' ] },
>         \   {'buns': ["``", "''"], 'expand_range': 0, 'input': [ '4' ], 'quoteescape': 1 }, 
>         \   {'buns': ['<<', '>>'], 'expand_range': 0, 'input': [ '5' ] },
>         \ ]
>         let g:sandwich#recipes = deepcopy(g:sandwich#default_recipes)+g:sandwich#recipes
>
> The sandwicd-add works as expected, vib works as well, but I can not make use of replace/delete functions by sd1/sd2,sr12 etc. Am I still missing something?

It seems good, I'm not sure why the keypresses `sd1`, `sd2` and  `sr12` do not work... It might be a problem related to unicode strings, is the keypresses `sd4` or `sd5` working? Anyway I will check my code.

**Comment by maxfl** [comment]
@machakann, thank you, looks amazing? I will try everything you wrote.

Yes, ``|'' and <<|>> work correctly.

**Comment by machakann** [comment]
Thanks! I will check.

**Comment by machakann** [comment]
I have finally found the reason, but it might be a bug of vim... I will ask persons much more familiar with vim about this problem. If needed I will report it to vim-dev.

Anyway I add a workaround for unicode characters (ea2d7e337ee628e926dfa25a9d7497ee29a6f63a). Now it works as you expected, please update and try.

**Comment by maxfl** [comment]
Thank you, it works amazing. I will try to switch to sandwich and will write in case there are any other problems.

**Comment by machakann** [comment]
Thank you. I will always welcome.

**Comment by maxfl** [comment]
Just in case, here is what I came up with for the simple sandwiches. The case of latex command/enviromnent is more complex, I need to think about it.
```vim
        let g:sandwich#recipes = [
        \	{'buns': ["`",                "'"],                 'nesting': 1, 'input': [ "u'" ]  },
        \	{'buns': ['“',                '”'],                 'nesting': 1, 'input': [ 'u"' ]  },
        \	{'buns': ['„',                '“'],                 'nesting': 1, 'input': [ 'U"', 'ug' ]  },
        \	{'buns': ['«',                '»'],                 'nesting': 1, 'input': [ 'u<', 'uf' ]  },
        \   {'buns': ["'''",              "'''"],               'nesting': 0, 'input': [ "3'" ], 'filetype': ['python'] },
        \   {'buns': ['"""',              '"""'],               'nesting': 0, 'input': [ '3"' ], 'filetype': ['python'] },
        \	{'buns': ["`",                "'"],                 'nesting': 1, 'input': [ "l'", "l`" ], 'filetype': ['tex', 'plaintex'] },
        \	{'buns': ["``",               "''"],                'nesting': 1, 'input': [ 'l"' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['"`',               "\"'"],               'nesting': 1, 'input': [ 'L"' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['<<',               '>>'],                'nesting': 1, 'input': [ 'l<' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left(',           '\right)'],           'nesting': 1, 'input': [ 'm(' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left[',           '\right]'],           'nesting': 1, 'input': [ 'm[' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left|',           '\right|'],           'nesting': 1, 'input': [ 'm|' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left\{',          '\right\}'],          'nesting': 1, 'input': [ 'm{' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left\langle ',    '\right\rangle '],    'nesting': 1, 'input': [ 'm<' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left(',           '\right)'],           'nesting': 1, 'input': [ 'm(' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left[',           '\right]'],           'nesting': 1, 'input': [ 'm[' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left|',           '\right|'],           'nesting': 1, 'input': [ 'm|' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left\{',          '\right\}'],          'nesting': 1, 'input': [ 'm{' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\left\langle ',    '\right\rangle '],    'nesting': 1, 'input': [ 'm<' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft(',        '\bigright)'],        'nesting': 1, 'input': [ 'M(' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft[',        '\bigright]'],        'nesting': 1, 'input': [ 'M[' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft|',        '\bigright|'],        'nesting': 1, 'input': [ 'M|' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft\{',       '\bigright\}'],       'nesting': 1, 'input': [ 'M{' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft\langle ', '\bigright\rangle '], 'nesting': 1, 'input': [ 'M<' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft(',        '\bigright)'],        'nesting': 1, 'input': [ 'M(' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft[',        '\bigright]'],        'nesting': 1, 'input': [ 'M[' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft|',        '\bigright|'],        'nesting': 1, 'input': [ 'M|' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft\{',       '\bigright\}'],       'nesting': 1, 'input': [ 'M{' ],    'filetype': ['tex',       'plaintex'] },
        \	{'buns': ['\bigleft\langle ', '\bigright\rangle '], 'nesting': 1, 'input': [ 'M<' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\begingroup ',     '\endgroup'],         'nesting': 0, 'input': [ 'lgr' ],   'filetype': ['tex',       'plaintex'], 'motionwise': ['char',  'block'] },
        \   {'buns': ['\begingroup',      '\endgroup'],         'nesting': 0, 'input': [ 'lgr' ],   'filetype': ['tex',       'plaintex'], 'motionwise': ['line'] },
        \ ]
```

**Comment by machakann** [comment]
It looks nice :)

**Comment by machakann** [comment]
Excuse me, now I find that there are several duplicates in your g:sandwich#recipes. It never makes problem, but it could be simplified. Isn't it?

```vim
        let g:sandwich#recipes = [
        \   {'buns': ["`",                "'"],                 'nesting': 1, 'input': [ "u'" ]  },
        \   {'buns': ['“',                '”'],                 'nesting': 1, 'input': [ 'u"' ]  },
        \   {'buns': ['„',                '“'],                 'nesting': 1, 'input': [ 'U"', 'ug' ]  },
        \   {'buns': ['«',                '»'],                 'nesting': 1, 'input': [ 'u<', 'uf' ]  },
        \   {'buns': ["'''",              "'''"],               'nesting': 0, 'input': [ "3'" ], 'filetype': ['python'] },
        \   {'buns': ['"""',              '"""'],               'nesting': 0, 'input': [ '3"' ], 'filetype': ['python'] },
        \   {'buns': ["`",                "'"],                 'nesting': 1, 'input': [ "l'", "l`" ], 'filetype': ['tex', 'plaintex'] },
        \   {'buns': ["``",               "''"],                'nesting': 1, 'input': [ 'l"' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['"`',               "\"'"],               'nesting': 1, 'input': [ 'L"' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['<<',               '>>'],                'nesting': 1, 'input': [ 'l<' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\left(',           '\right)'],           'nesting': 1, 'input': [ 'm(' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\left[',           '\right]'],           'nesting': 1, 'input': [ 'm[' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\left|',           '\right|'],           'nesting': 1, 'input': [ 'm|' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\left\{',          '\right\}'],          'nesting': 1, 'input': [ 'm{' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\left\langle ',    '\right\rangle '],    'nesting': 1, 'input': [ 'm<' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\bigleft(',        '\bigright)'],        'nesting': 1, 'input': [ 'M(' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\bigleft[',        '\bigright]'],        'nesting': 1, 'input': [ 'M[' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\bigleft|',        '\bigright|'],        'nesting': 1, 'input': [ 'M|' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\bigleft\{',       '\bigright\}'],       'nesting': 1, 'input': [ 'M{' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\bigleft\langle ', '\bigright\rangle '], 'nesting': 1, 'input': [ 'M<' ],    'filetype': ['tex',       'plaintex'] },
        \   {'buns': ['\begingroup ',     '\endgroup'],         'nesting': 0, 'input': [ 'lgr' ],   'filetype': ['tex',       'plaintex'], 'motionwise': ['char',  'block'] },
        \   {'buns': ['\begingroup',      '\endgroup'],         'nesting': 0, 'input': [ 'lgr' ],   'filetype': ['tex',       'plaintex'], 'motionwise': ['line'] },
        \ ]
```

**Comment by maxfl** [comment]
Indeed. Thanks for pointing this out.

**Comment by machakann** [comment]
Just for my note: The bug related with multibyte characters has been fixed at the patch 7.4.771. Thank Mr. Nakadaira!

https://groups.google.com/forum/#!msg/vim_dev/XPY4cZaKUag/2jS8NAJvYeoJ
https://groups.google.com/forum/#!msg/vim_dev/JckBTr7CpDY/AyouJCxJJAoJ
